Upgrade to more recent .NET SDK

dougbu opened the pull request.

dougbu committed:
Update SDK to preview.5.20251.2
- includes support for the `net5.0` TFM
- includes fix for `$(PublishDepsFilePath)`
- includes fix for incorrect NU5118 warnings when `'$(PackAsTool)' == 'true'`

dougbu commented:
/fyi @ViktorHofer @lukas-lansky @riarenas

dougbu commented:
I believe the new SDK brings along a number of breaking changes. New errors and warnings include
```
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-dev-certs\Release\netcoreapp5.0\publish\dotnet-dev-certs.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-dev-certs.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-dev-certs\src\dotnet-dev-certs.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-dev-certs\Release\netcoreapp5.0\publish\dotnet-dev-certs.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-dev-certs.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-dev-certs\src\dotnet-dev-certs.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-sql-cache\Release\netcoreapp5.0\publish\dotnet-sql-cache.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-sql-cache.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-sql-cache\src\dotnet-sql-cache.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-user-secrets\Release\netcoreapp5.0\publish\dotnet-user-secrets.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-user-secrets.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-user-secrets\src\dotnet-user-secrets.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-user-secrets\Release\netcoreapp5.0\publish\dotnet-user-secrets.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-user-secrets.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-user-secrets\src\dotnet-user-secrets.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-sql-cache\Release\netcoreapp5.0\publish\dotnet-sql-cache.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-sql-cache.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-sql-cache\src\dotnet-sql-cache.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-watch\Release\netcoreapp5.0\publish\dotnet-watch.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-watch.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-watch\src\dotnet-watch.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\dotnet-watch\Release\netcoreapp5.0\publish\dotnet-watch.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-watch.deps.json' [F:\workspace\_work\1\s\src\Tools\dotnet-watch\src\dotnet-watch.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\Microsoft.dotnet-openapi\Release\netcoreapp5.0\publish\dotnet-openapi.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-openapi.deps.json' [F:\workspace\_work\1\s\src\Tools\Microsoft.dotnet-openapi\src\Microsoft.dotnet-openapi.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5118: File 'F:\workspace\_work\1\s\artifacts\bin\Microsoft.dotnet-openapi\Release\netcoreapp5.0\publish\dotnet-openapi.deps.json' is not added because the package already contains file 'tools\netcoreapp5.0\any\dotnet-openapi.deps.json' [F:\workspace\_work\1\s\src\Tools\Microsoft.dotnet-openapi\src\Microsoft.dotnet-openapi.csproj]
...
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5129: - At least one .targets file was found in 'buildTransitive/netcoreapp5.0/', but 'buildTransitive/net5.0/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.targets' was not. [F:\workspace\_work\1\s\src\Mvc\Mvc.Razor.RuntimeCompilation\src\Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj]
...
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5129: - At least one .targets file was found in 'buildTransitive/netcoreapp5.0/', but 'buildTransitive/net5.0/Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.targets' was not. [F:\workspace\_work\1\s\src\Mvc\Mvc.Razor.RuntimeCompilation\src\Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.csproj]
...
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.4.20202.8\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5129: - At least one .targets file was found in 'build/netcoreapp5.0/', but 'build/net5.0/Microsoft.AspNetCore.Mvc.Testing.targets' was not. [F:\workspace\_work\1\s\src\Mvc\Mvc.Testing\src\Microsoft.AspNetCore.Mvc.Testing.csproj]
...
Unhandled exception. System.IO.DirectoryNotFoundException: Could not find a part of the path 'F:\workspace\_work\1\s\artifacts\obj\InProcessWebSite\Release\netcoreapp5.0\win-x64\x64\aspnetcorev2_inprocess.dll'.
     at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
     at System.IO.File.Copy(String sourceFileName, String destFileName, Boolean overwrite)
     at TestTasks.InjectRequestHandler.Main(String[] args) in F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\TestTasks\InjectRequestHandler.cs:line 61
F:\workspace\_work\1\s\src\Servers\IIS\build\testsite.props(73,5): error MSB3073: The command "dotnet F:\workspace\_work\1\s\src\Servers\IIS\build\..\IIS\test\testassets\TestTasks\bin\Release\netcoreapp5.0\TestTasks.dll "F:\workspace\_work\1\s\artifacts\obj\InProcessWebSite\Release\netcoreapp5.0\win-x64\InProcessWebSite.deps.json" win-x64 " exited with code -532462766. [F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\InProcessWebSite\InProcessWebSite.csproj]
    24 Warning(s)
    1 Error(s)
```

The final error implies native assets aren't being copied where they're expected.

So, @ViktorHofer @dsplaisted @rainersigwald @rrelyea what's changed and how do we need to react in this repo?

rrelyea commented:
- nu5118 - I saw something like NU5118 when i was testing multitargetting to net5.0 and netcoreapp5.0 -- and nuget thinks of them as equivalent.  pick one.
- nu5129 - https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5129

dougbu commented:
@rrelyea
- We're not using the `net5.0` TFM at all in this repo (yet). The projects where NU5118 warnings occur all use `<PackAsTool>true</PackAsTool>`.
- There's no misalignment between the package IDs and the "missing" *.targets files. It's much more like NuGet or the .NET SDK no longer support the `netcoreapp5.0` TFM.

FYI we don't do much more than the following in the projects showing NU5129 warnings.
``` msbuild
  <ItemGroup>
    <None Include="targets\$(PackageId).targets" Pack="true" PackagePath="build\$(DefaultNetCoreTargetFramework)\$(Packa
geId).targets" />
    <None Include="targets\$(PackageId).targets" Pack="true" PackagePath="buildTransitive\$(DefaultNetCoreTargetFramewor
k)\$(PackageId).targets" />
  </ItemGroup>
```

Are you saying the fix is to change our `$(DefaultNetCoreTargetFramework)` property to `net5.0` and workaround the fact Microsoft.NetCore.App still uses the `netcoreapp5.0` TFM? (That sounds a bit scary.)

rrelyea commented:
what build of sdk were you using before? what version are you upgrading to?

where you see us treating netcoreapp5.0 as net5.0, and you are running into problems...
the right long term thing to do is move to net5.0.

until then, you may be able to live with ignoring the warning (via nowarn)?
looking at previous packages that were built before this happened may help us determine what files the output package is created with.

@zkat

dougbu commented:
@rrelyea this PR makes the following change
``` diff
-    "version": "5.0.100-preview.2.20120.3"
+    "version": "5.0.100-preview.4.20202.8"
```

dougbu commented:
> until then, you may be able to live with ignoring the warning (via nowarn)?

That won't solve the error shown above.

rrelyea commented:
by error, you mean the exception?

we generally tried to leave netcoreapp5.0 working in parallel with net5.0.
seems like these projects may be problematic.
that exception looks like it is looking for a file (with netcoreapp5.0 in the path).
most likely, that file is next to it (with net5.0 in the path).

so you need to figure out if you can move forward, or if you need to stay back.
if you have to stay back, you might need to make your test understand the hybrid state you are in.

when does Microsoft.NetCore.App get fixed and by who?

dougbu commented:
@rrelyea why would `$(IntermediateOutputPath)` be `artifacts\obj\InProcessWebSite\Release\net5.0\win-x64\x64\aspnetcorev2_inprocess.dll` when the `$(OutputPath)` is `bin\Release\netcoreapp5.0\InProcessWebSite.dll`? (I'm adding more logging to the failing `InjectRequestHandler` class.)

> when does Microsoft.NetCore.App get fixed and by who?

@ericstj @ViktorHofer I don't see any mention of `net5.0` in dotnet/runtime yet. When are you planning to make the move?

If that's not coming soon, we can't change our default TFM until we get VS16.6 on the agents or I resolve the problems seen in #20748. The short-term workaround would be to update `InjectRequestHandler` and ignore the new NuGet warnings.

dougbu committed:
!!temporary!! Suppress NU5129 warnings
- build/ and buildTransitive/ naming conventions do not treat TFMs as equivalent

ViktorHofer commented:
We will switch to net5.0 on May 4th: https://github.com/dotnet/runtime/issues/35202

dougbu commented:
My added logging is dumping into `/dev/null` at the moment -- forgot `stdout` is mostly lost in our test frameworks. Working on local reproduction to figure out where the in-proc handler actually lands.

dougbu commented:
There are a couple of things going on here:
1. With the new SDK, `restore` operations frequently hit the following error:
    ```
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(128,5): error : Cannot create a file when that file already exists. [C:\dd\dnx\aspnetcore\src\Servers\IIS\IISIntegration\test\Tests\Microsoft.AspNetCore.Server.IISIntegration.Tests.csproj]
    ```
    Unfortunately there is no information in the binary log about _which_ file already exists. Beyond seeing multiple `msbuild` processes restoring the same project e.g. Microsoft.AspNetCore.Testing, the symptoms aren't clear.

    @rainersigwald and @rrelyea are the build / restore parallelism issues an expected regression? Are there workarounds or ways to debug the issue?

2. `InjectRequestHandler` in TestTasks.dll is given `$(PublishDepsFilePath)` as one of its arguments and, for the failing command, that property value has changed from
    ```
    bin\Release\netcoreapp5.0\publish\\InProcessWebSite-Standalone-x64\InProcessWebSite.deps.json
    ```
    in 'master' to the following in this branch
    ```
    artifacts\obj\InProcessWebSite\Release\netcoreapp5.0\win-x64\InProcessWebSite.deps.json
    ```
    The value change doesn't matter when the command is also given an empty `$(RuntimeIdentifier)`. But, when the second argument is non-empty, `InjectRequestHandler` attempts to copy aspnetcorev2_inprocess.dll within the (supposed) publish/ directory. Since the command is actually given a *.deps.json path within obj/ and aspnetcorev2_inprocess.dll is only copied into bin/ and publish/ directories, the copy fails.

    The root cause of the problem is a breaking change in Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets, from the following in 5.0.100-preview.2.20120.3
    ``` msbuild
      <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == '' And '$(PublishSingleFile)' != 'true'">$(PublishDir)$(ProjectDepsFileName)</PublishDepsFilePath>
      <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == '' And '$(PublishSingleFile)' == 'true'">$(IntermediateOutputPath)$(ProjectDepsFileName)</PublishDepsFilePath>
    ```
    to the following in 5.0.100-preview.4.20202.8
    ``` msbuild
    <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == ''">$(IntermediateOutputPath)$(ProjectDepsFileName)</PublishDepsFilePath>
    ```
    (The `GeneratePublishDependencyFile` target has also been removed or renamed.)

    @swaroop-sridhar and @dsplaisted was the breaking change in https://github.com/dotnet/sdk/commit/1f2a4a110e356808d721358ce358de4f88923795#diff-7dc833d02f198e5a9f0ab88e229db18aL996-R980 intended? We might be able to work around the issue but I'd rather not if it'll be fixed soon. (Since dotnet/sdk#10740 was about single-file publish operations, I'm surprised the behaviour of regular publish ops changed at all.)

swaroop-sridhar commented:
> Was the breaking change in dotnet/sdk@1f2a4a1#diff-7dc833d02f198e5a9f0ab88e229db18aL996-R980 intended? We might be able to work around the issue but I'd rather not if it'll be fixed soon. (Since dotnet/sdk#10740 was about single-file publish operations, I'm surprised the behaviour of regular publish ops changed at all.)

@dougbu  The change to generate the `deps.json` file in the `obj` directory for all builds was intentional -- because it simplifies the `ResolvedFileToPublish` computation and ordering of certain publish targets.

However, the definition of `PublishDepsFilePath` could be left unchanged (for non-single-file cases).
For example, I can change the definition such that `PublishDepsFilePath` is:
* `$(PublishDir)$(ProjectDepsFileName)` for normal publish
* Empty for single-file publish.

You could use `$(PublishDir)$(AssemblyName).deps.json`  in the test code to circumvent the problem in the meantime.

dougbu commented:
> You could use `$(PublishDir)$(AssemblyName).deps.json` in the test code to circumvent the problem in the meantime.

We also need to edit the .deps file. adding information about a native assembly. That's why our command executes after `GeneratePublishDependencyFile`. Does `$(PublishDir)$(AssemblyName).deps.json` exist once `GeneratePublishDependencyFile` is done?

Since simplifying `InjectRequestHandler` to avoid the copy didn't work, I'm instead going to look at passing the `$(PublishDir)` directory on the command line.

dougbu commented:
But,
> I can change the definition such that `PublishDepsFilePath` is

sounds better to me.

dougbu commented:
&hellip; because that restores the previous values and customers may also be depending on it.

ViktorHofer commented:
@dougbu please update to `5.0.100-preview.5.20227.9` which contains a fix for a runtime regression that we had in for ~20 days: https://github.com/dotnet/runtime/commit/a5414a2bdf5a3e45e3800530bfd35de128971ac1

ViktorHofer commented:
> @rainersigwald and @rrelyea are the build / restore parallelism issues an expected regression? Are there workarounds or ways to debug the issue?

I believe restore parallelism was never supported? At least not for common dependency nodes. cc @nkolev92

dougbu commented:
> I believe restore parallelism was never supported?

I was talking about the usual `msbuild` approach in which different parts of a multi-project restore operation run in separate processes.

dougbu commented:
>@dougbu please update to `5.0.100-preview.5.20227.9`

We don't use the SDK to determine the version of Microsoft.NetCore.App we build against; doing so would require a loop through the dependency graph and back to dotnet/aspnetcore before we'd see a dotnet/runtime fix. But, I can move to the latest known good SDK anyhow. If you want us to move to more recent dotnet/runtime versions, regular Maestro++ updates such as #21238 are the place to look.

@dsplaisted I see Microsoft.Dotnet.Sdk.Internal v5.0.100-preview.5.20227.9 is available in the dotnet5-transport feed. Is that version not only the latest but also the greatest?

rainersigwald commented:
> I was talking about the usual `msbuild` approach in which different parts of a multi-project restore operation run in separate processes.

Should still be ok. Got a binlog of a failure, @dougbu?

dougbu commented:
Sorry for the delay; had to reboot (for other reasons) then reproduce the failure to grab a binary log. The command I executed in this branch was:
```
.\build.cmd -Configuration Release -buildNative -projects .\src\Servers\IIS\**\*.*proj -bl
```

The file should soon be in my \\GarnetB\Share folder (still uploading). If you need it shared somewhere else e.g. OneDrive, please let me know.

swaroop-sridhar commented:
> But,
> 
> > I can change the definition such that `PublishDepsFilePath` is
> 
> sounds better to me.

OK I'll fix it now. thanks.

dougbu commented:
@rainersigwald I reproduced the wrong problem, trying a few more things&hellip;

swaroop-sridhar commented:
> > But,
> > > I can change the definition such that `PublishDepsFilePath` is
> > 
> > 
> > sounds better to me.
> 
> OK I'll fix it now. thanks.

https://github.com/dotnet/sdk/pull/11462

dougbu commented:
@rainersigwald two binary logs are finally available.

Apparently it helps _not_ to tie up a `msbuild` process with native compilation and, still, I need to hit on the right number of projects to pass to the repo-level build. E.g. the following tends to hit 3 failures or none:
``` powershell
.\build.cmd -ci -Configuration Release -arch x64 -projects .\src\Servers\IIS\IIS\test\**\*.*proj -bl
```
And, the following usually hits one failure:
``` powershell
.\build.cmd -ci -Configuration Release -arch x64 -projects .\src\Servers\IIS\IIS\test\testassets\**\*.*proj -bl
```

rainersigwald commented:
Looked at `Build.test.binlog`. The way the build script is structured, you're calling the `Restore` target on a bunch of projects in parallel. That's not supported. NuGet supports solution-level restore with parallelism by walking the project graph in parallel, then restoring in a single operation in the solution. So I think @ViktorHofer nailed it.

dougbu commented:
> you're calling the Restore target on a bunch of projects in parallel. That's not supported

I can't call a solution build a fix because such builds do not support any concepts of optionality. Further, eng/build.props is built into Arcade. Changing one of the core Arcade conventions is scarier to me than running an initial `restore` with `-maxCpuCount:1`.

ViktorHofer commented:
Doug I can help you with that. Look at how dotnet/runtime implemented a single restore/build/test entry point which is msbuild driven, means you will keep the current flexibility that you have with ProjectToBuild:

1. Arcade configuration setting: https://github.com/dotnet/runtime/blob/f185bd5571f2500b4843d145418c4418747246b6/eng/Build.props#L6
2. Single entry point msbuild file: https://github.com/dotnet/runtime/blob/master/Build.proj#L1-L27 (Ignore everything below)
3. Subsets.props which defines all the projects to build and allows filtering based on the /p:Subset property: https://github.com/dotnet/runtime/blob/master/eng/Subsets.props

This also reduces the restore times as nuget can be smarter about what and how often to restore with the complete dependency graph.

dougbu commented:
> Look at how dotnet/runtime implemented a single restore/build/test entry point which is msbuild driven

I've seen this and it looks like an end run around the Arcade conventions. If Arcade isn't using `msbuild` correctly, I'd rather we fix it than layer workarounds on top of it. /cc @markwilkie 

> This also reduces the restore times as nuget can be smarter about what and how often to restore with the complete dependency graph.

Does this matter on the CI or is this part of your comment specifically about incremental builds on a 
dev's machine?

dougbu commented:
@markwilkie to clarify: I'm wondering about a feature in Arcade that would allow us to more easily opt into having a single top-level project or solution when doing repo builds.

Either supporting an eng/Build.proj file (which we'd populate w/ lots of conditional `@(ProjectReference)` items) or writing out a solution file would likely work. The preference I expressed above is about dotnet/aspnetcore avoiding boilerplate other repos would likely want, given we're currently doing an unsupported gesture [in Arcade's Build.proj](https://github.com/dotnet/arcade/blob/592ec36c7fcfbd4c3a6025c61456e351aa2c9057/src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj#L222-L228) when `@(ProjectToBuild)` contains more than one item.

markwilkie commented:
Sure - I'd be open to looking more at taking a PR to arcade with the single project entry point.  This is especially true if more than one repo needs this functionality.  The one thing we'd want to watch out for is unnecessary complexity...

dougbu commented:
/fyi @swaroop-sridhar got his fix in before I pushed a workaround attempt. Now waiting for https://dev.azure.com/dnceng/internal/_build/results?buildId=624059 and whatever needs to follow that before the new SDK is available.

dougbu commented:
> This is especially true if more than one repo needs this functionality.

In a quick search, I found most ASP.NET Core repos, dotnet/roslyn-sdk, dotnet/spark, dotnet/test-templates, and dotnet/windowsdesktop all use this feature. Arcade also tests multiple `@(ProjectToBuild)` items in https://github.com/dotnet/arcade/blob/f657be5cb7cd4920334dd9162173b131211a1e17/src/Microsoft.DotNet.Arcade.Sdk.Tests/testassets/RepoWithConditionalProjectsToBuild/eng/Build.props

I didn't check to see how many other repos have no eng/Build.props file and instead place more than one .sln file in the repo root (Arcade's fallback for `@(ProjectToBuild)`).

ViktorHofer commented:
> I've seen this and it looks like an end run around the Arcade conventions. If Arcade isn't using msbuild correctly, I'd rather we fix it than layer workarounds on top of it. /cc @markwilkie

I have to disagree. Arcade was initially designed with the assumption that repositories expose a single entrypoint which Arcade's Build.proj file can invoke. That was preferably a Solution File and in cases where this wasn't possible (corefx) we asked the Team to add an extension point (Build.props) which allows to define the project which should get build. It was @natemcmaster who such an extension point: https://github.com/dotnet/arcade/commit/f657be5cb7cd4920334dd9162173b131211a1e17.

It was @natemcmaster as well who later then added the ability to specify a custom Restore target: https://github.com/dotnet/arcade/commit/b5016f5688dc8ca9f3e4811ee7e2e86ad8907a40. He also added a workaround to avoid NuGet restore conflicts when invoking Restore in parallel: https://github.com/NuGet/Home/issues/7648.

I added the `Build.proj` file in dotnet/runtime for several reasons:
- Allows to call `dotnet restore`, `dotnet build` and `dotnet test` from the repository root even if the SDK hasn't been restored yet (ie a fresh repo). This is possible because we have two scripts called `dotnet.cmd/sh` in the repo root that handle the SDK bootstrapping: https://github.com/dotnet/runtime/blob/master/dotnet.cmd. That's the corresponding ask: https://github.com/dotnet/runtime/issues/947. My intent is that will meet with @markwilkie and his team at some point to discuss how to further support the non-script based entry point by sequencing stuff that's currently happening inside Arcade's Build.proj via an SDK or easy imports into the repository's entry point.
- Parallel restore isn't supported by NuGet when having interleaving child nodes. Using a single restore entry point and restoring sequentially would be terribly slow for repositories with many projects (ie dotnet/runtime, aspnetcore). Hence the NuGet team added the new [static graph restore mode](https://docs.microsoft.com/en-us/nuget/release-notes/nuget-5.5) which reduces the no-op restore drastically (up to 90%) and also improves clean restore because of a single restore entry point.
My plan here is to meet with @markwilkie and team as well to discuss how we can support that feature in all arcadified repositories as non-disruptively as possible.

We wouldn't have this discussion if Solution Files would allow globbing. Hence the obvious alternative is https://github.com/microsoft/MSBuildSdks/tree/master/src/Traversal which we use in dotnet/runtime: https://github.com/dotnet/runtime/blob/0764d8c11760069056c3f2c4509644c161fce490/Build.proj#L1.

ViktorHofer commented:
cc @jeffkl @tmat @danmosemsft

tmat commented:
I like @ViktorHofer's approach. `dotnet.cmd` in the root and a single entry point to build everything (`build.proj` or a solution).
I also agree that ideally VS would support opening build.proj that uses Microsoft.Build.Traversal like a solution.

jeffkl commented:
Side note: You can generate a solution on the fly with my [SlnGen tool](https://github.com/microsoft/slngen).  It works with `Microsoft.Build.Traversal` and as long as you declare `ProjectReference` items it'll generate a solution at any level of the tree.  In parallel I'm trying to work with the VS team to see if we can just make Visual Studio load projects transitively when you open them and essentially generate a solution on the fly.

tmat commented:
Yes, you can use slngen  but i consider it a work around.

tmat commented:
@jeffkl 
> I'm trying to work with the VS team to see if we can just make Visual Studio load projects transitively when you open them and essentially generate a solution on the fly.

That would be awesome!

dougbu commented:
Generalizing the bootstrap infrastructure as dotnet/runtime has done in the dotnet.* files and as dotnet/aspnetcore and related repos have done in build.* files sounds great to me. Microsoft.Build.Traversal also sounds like a good idea, especially since it should ease managing `BlahInParallel` choices.

@ViktorHofer I'm not quite sure what part you're disagreeing with me. My suggestion was intended to be along the lines of "let's generalize what dotnet/runtime is doing". Could you clarify?

ViktorHofer commented:
@dougbu this comment made me assume that you are against the Build.proj approach and instead want to improve Arcade's Build.proj (which I believe isn't the right path forward):

> I've seen this and it looks like an end run around the Arcade conventions. If Arcade isn't using msbuild correctly, I'd rather we fix it than layer workarounds on top of it.

I missed your other comments which clarifies your comment:

> to clarify: I'm wondering about a feature in Arcade that would allow us to more easily opt into having a single top-level project or solution when doing repo builds.

Sounds like we are all on the same boat which is fantastic 😀. I'll create corresponding issues in Arcade to follow-up on the discussion points.

@dougbu I believe we should get an upated SDK out later today if the dependency ingestion from sdk -> installer runs flawlessly 🤞. What is your plan to mitigate the parallel restore issues? Adding the Build.proj file similar to what we have in dotnet/runtime?

dougbu commented:
> What is your plan to mitigate the parallel restore issues?

Those issues don't affect our AzDO pipelines; I've only seen problems when building a smallish subset of the repo, likely because that increases the likelihood multiple `msbuild` instances will try to restore shared projects at once. So, I'm shelving problem 1 for the moment.

ViktorHofer commented:
Opened https://github.com/dotnet/arcade/issues/5384 to track the Build.proj entrypoint discussion.

ViktorHofer commented:
@dougbu  I fixed the issue in dotnet/sdk that you were seeing (https://github.com/dotnet/sdk/issues/11521) which caused the NU5118 errors and just upgraded the SDK being used by this PR.

dougbu commented:
Thanks @ViktorHofer❕ I need to revert a couple of earlier commits and the NU5118 suppressions. Will push that shortly&hellip;

@rrelyea am I correct newer SDKs will still produce incorrect NU5129 warnings about "missing" `net5.0` files under build/ or buildTransitive/? (The issue is the naming conventions don't treat `net5.0` and `netcoreapp5.0` as equivalent.)

dougbu commented:
:up::date: to rebase and to suppress _only_ NU5129 warnings

rrelyea commented:
I don't believe we've done any work to make NU5129 understand when a package is half net5.0 and half netcoreapp5.0

dougbu commented:
/fyi @rrelyea in this case, the package is entirely `netcoreapp5.0`, not "half and half".

dougbu commented:
@swaroop-sridhar the latest SDK seems to contain a new breaking change. Errors start at https://dev.azure.com/dnceng/public/_build/results?buildId=626806&view=logs&j=1b89928a-2219-5ef9-602f-f95beb3da4dc&t=bcce746a-741e-59b7-adec-a4a82d8c5e52&l=1652 Why doesn't the .deps file exist in a target that runs after `GeneratePublishDependencyFile`? Do we need to update the [targets here](https://github.com/dotnet/aspnetcore/blob/ba4dc9f347658975e6c9aa2a93b22fef3193681e/src/Servers/IIS/build/testsite.props#L62-L74)?
```
  Unhandled exception. System.IO.FileNotFoundException: Could not find file 'F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\InProcessWebSite\bin\Release\netcoreapp5.0\publish\InProcessWebSite-Portable\InProcessWebSite.deps.json'.
  File name: 'F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\InProcessWebSite\bin\Release\netcoreapp5.0\publish\InProcessWebSite-Portable\InProcessWebSite.deps.json'
     at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
     at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
     at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
     at System.IO.StreamReader.ValidateArgsAndOpenPath(String path, Encoding encoding, Int32 bufferSize)
     at System.IO.StreamReader..ctor(String path)
     at System.IO.File.OpenText(String path)
     at TestTasks.InjectRequestHandler.Main(String[] args) in F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\TestTasks\InjectRequestHandler.cs:line 27
...
F:\workspace\_work\1\s\src\Servers\IIS\build\testsite.props(73,5): error MSB3073: The command "dotnet F:\workspace\_work\1\s\src\Servers\IIS\build\..\IIS\test\testassets\TestTasks\bin\Release\netcoreapp5.0\TestTasks.dll "bin\Release\netcoreapp5.0\publish\\InProcessWebSite-Portable\InProcessWebSite.deps.json"  " exited with code -532462766. [F:\workspace\_work\1\s\src\Servers\IIS\IIS\test\testassets\InProcessWebSite\InProcessWebSite.csproj]
```

dougbu commented:
@dsplaisted, @swaroop-sridhar the problem is that https://github.com/dotnet/sdk/blob/c09668ef26a1c8917144e897b2989524cc332d47/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L1006-L1008 now generates the .deps file in `$(IntermediateOutputPath)`. I also don't see where `$(IntermediateDepsFilePath)` would be copied to `$(PublishDepsFilePath)`. If there's already a fix in progress, please let me know.

dougbu commented:
/fyi dotnet/sdk#11527

rrelyea commented:
May is here ... maybe you'll be able to switch to net5.0 soon? https://github.com/dotnet/runtime/issues/35202

swaroop-sridhar commented:
@dougbu, in this case, I think it is better for `InjectRequestHandlerOnPublish` to run after `ComputeAndCopyFilesToPublishDirectory` which will copy the deps file from `IntermediateDepsFilePath` to `PublishDepsFilePath`.

dougbu committed:
Run `InjectRequestHandlerOnPublish` later
- late enough for .deps file to exist i.e. after the `CopyFilesToPublishDirectory` target
- `AfterTargets` is a bit flaky; use `BeforeTargets` with empty targets instead

dougbu commented:
Seems we have a new, unrelated break. For now at least, I'm removing the config file that causes `dotnet-serve` to be restored. Any objection @pranavkm❔
```
Downloading vswhere
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin\msbuild.exe /nologo /bl:F:\workspace\_work\1\s\artifacts\log\Release\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /m /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=false /p:__ToolsetLocationOutputFile=F:\workspace\_work\1\s\artifacts\toolset\5.0.0-beta.20221.14.txt /t:__WriteToolsetLocation /v:minimal F:\workspace\_work\1\s\artifacts\toolset\restore.proj
  System.AggregateException: One or more errors occurred. (more than one row for dotnet-serve-1.5.0-net5.0-any-dotnet-serve)
   ---> Microsoft.DotNet.ToolPackage.ResolverCacheInconsistentException: more than one row for dotnet-serve-1.5.0-net5.0-any-dotnet-serve
     at Microsoft.DotNet.ToolPackage.LocalToolsResolverCache.TryGetMatchingRestoredCommand(RestoredCommandIdentifier restoredCommandIdentifier, CacheRow[] cacheTable, RestoredCommand& restoredCommandList)
     at Microsoft.DotNet.ToolPackage.LocalToolsResolverCache.TryLoad(RestoredCommandIdentifier restoredCommandIdentifier, RestoredCommand& restoredCommand)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.PackageHasBeenRestored(ToolManifestPackage package, String targetFramework)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.<>c__DisplayClass11_0.<Execute>b__0(ToolManifestPackage package)
     at System.Linq.Parallel.SelectQueryOperator`2.SelectQueryOperatorResults.GetElement(Int32 index)
     at System.Linq.Parallel.QueryResults`1.get_Item(Int32 index)
     at System.Linq.Parallel.PartitionedDataSource`1.ListContiguousIndexRangeEnumerator.MoveNext(T& currentElement, Int32& currentKey)
     at System.Linq.Parallel.StopAndGoSpoolingTask`2.SpoolingWork()
     at System.Linq.Parallel.SpoolingTaskBase.Work()
     at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
     at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
     at System.Threading.Tasks.Task.InnerInvoke()
     at System.Threading.Tasks.Task.<>c.<.cctor>b__276_0(Object obj)
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
  --- End of stack trace from previous location ---
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
     --- End of inner exception stack trace ---
     at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
     at System.Linq.Parallel.SpoolingTask.SpoolStopAndGo[TInputOutput,TIgnoreKey](QueryTaskGroupState groupState, PartitionedStream`2 partitions, SynchronousChannel`1[] channels, TaskScheduler taskScheduler)
     at System.Linq.Parallel.DefaultMergeHelper`2.System.Linq.Parallel.IMergeHelper<TInputOutput>.Execute()
     at System.Linq.Parallel.MergeExecutor`1.Execute()
     at System.Linq.Parallel.MergeExecutor`1.Execute[TKey](PartitionedStream`2 partitions, Boolean ignoreOutput, ParallelMergeOptions options, TaskScheduler taskScheduler, Boolean isOrdered, CancellationState cancellationState, Int32 queryId)
     at System.Linq.Parallel.PartitionedStreamMerger`1.Receive[TKey](PartitionedStream`2 partitionedStream)
     at System.Linq.Parallel.UnaryQueryOperator`2.UnaryQueryOperatorResults.GivePartitionedStream(IPartitionedStreamRecipient`1 recipient)
     at System.Linq.Parallel.QueryOperator`1.ExecuteAndGetResultsAsArray()
     at System.Linq.ParallelEnumerable.ToArray[TSource](ParallelQuery`1 source)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.Execute()
     at Microsoft.DotNet.Cli.DotNetTopLevelCommandBase.RunCommand(String[] args)
     at Microsoft.DotNet.Tools.Tool.ToolCommand.Run(String[] args)
     at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
     at Microsoft.DotNet.Cli.Program.Main(String[] args)
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20221.14\tools\Tools.proj(44,5): error MSB3073: The command ""F:\workspace\_work\1\s\.dotnet\dotnet.exe" tool restore" exited with code 1.
##[error].packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20221.14\tools\Tools.proj(44,5): error MSB3073: The command ""F:\workspace\_work\1\s\.dotnet\dotnet.exe" tool restore" exited with code 1.
##[error](NETCORE_ENGINEERING_TELEMETRY=Build) Build failed.
```

dsplaisted commented:
@wli3 Do you know what would be causing this tool restore failure?

dougbu commented:
@chcosta and @Chrisboh see a lot of 529 errors in NuGetClient. Is there anything we can do besides retry when AzDO throttles our feeds❔
```
/datadisks/disk1/workspace/_work/1/s/.dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error : Failed to download package 'Microsoft.NETCore.App.Host.win-x64.5.0.0-preview.5.20230.9' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/flat2/microsoft.netcore.app.host.win-x64/5.0.0-preview.5.20230.9/microsoft.netcore.app.host.win-x64.5.0.0-preview.5.20230.9.nupkg'.
/datadisks/disk1/workspace/_work/1/s/.dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error : Response status code does not indicate success: 429 (Request was blocked due to exceeding usage of resource 'ATCPU' in namespace 'IPAddress'. For more information on why your request was blocked, see the topic "Rate limits" on the Microsoft Web site (https://go.microsoft.com/fwlink/?LinkId=823950). (DevOps Activity ID: 2210484D-052D-4AB8-B816-C54B7E296E40)).
##[error].dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error : Failed to download package 'Microsoft.NETCore.App.Host.win-x64.5.0.0-preview.5.20230.9' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/7d9f5c21-0d79-403f-bfe3-9a4506529760/nuget/v3/flat2/microsoft.netcore.app.host.win-x64/5.0.0-preview.5.20230.9/microsoft.netcore.app.host.win-x64.5.0.0-preview.5.20230.9.nupkg'.
Response status code does not indicate success: 429 (Request was blocked due to exceeding usage of resource 'ATCPU' in namespace 'IPAddress'. For more information on why your request was blocked, see the topic "Rate limits" on the Microsoft Web site (https://go.microsoft.com/fwlink/?LinkId=823950). (DevOps Activity ID: 2210484D-052D-4AB8-B816-C54B7E296E40)).
/datadisks/disk1/workspace/_work/1/s/.dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error : Central Directory corrupt.
/datadisks/disk1/workspace/_work/1/s/.dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error :   Invalid argument
##[error].dotnet/sdk/5.0.100-preview.5.20251.2/NuGet.targets(128,5): error : Central Directory corrupt.
```

dougbu commented:
/azp run

dougbu commented:
/ping @pranavkm 
> Seems we have a new, unrelated break. For now at least, I'm removing the config file that causes dotnet-serve to be restored. Any objection @pranavkm❔

That is, this PR removes .config/dotnet-tools.json and seems to build and test fine without it. I'd appreciate your confirmation `dotnet-serve` (the only configured tool) isn't needed anymore.

rrelyea commented:
@dougbu - forwarded you an email thread related to that tools problem.

dougbu commented:
/azp run

dougbu commented:
/azp run

wli3 commented:
Sorry for being late. Agree with the email thread. Setting `DOTNET_CLI_HOME` should solve this issue.

pranavkm commented:
> That is, this PR removes .config/dotnet-tools.json and seems to build and test fine without it. I'd appreciate your confirmation dotnet-serve (the only configured tool) isn't needed anymore.

@dougbu do you still need to remove the file? The blazor-wasm template tests were removed in master, but should be back in a couple of weeks.

ViktorHofer commented:
> @dougbu do you still need to remove the file? The blazor-wasm template tests were removed in master, but should be back in a couple of weeks.

We can fix the overall issue meanwhile in Arcade so that when the tests come back we can bring the global tool back as well.

dougbu commented:
@pranavkm do the blazor template tests still use `dotnet-serve`? If yes, I'll revert the .config/ file removal and deal with `$env:DOTNET_CLI_HOME` separately.

@ViktorHofer very few of the builds in this repo use the Arcade scripts. We'll need to do our own fixes here. Of course, an Arcade fix will help many others.

pranavkm commented:
> do the blazor template tests still use dotnet-serve

Yes, in the blazor-wasm branch. Our plan is to move these changes upstream in the next few weeks.

dougbu commented:
Quarantined template tests are failing consistently e.g.
```
    Templates.Test.RazorClassLibraryTemplateTest.RazorClassLibraryTemplate_WithViews_Async [FAIL]
      System.TypeInitializationException : The type initializer for 'Templates.Test.Helpers.TemplatePackageInstaller' threw an exception.
      ---- System.InvalidOperationException : Sequence contains no matching element
      Stack Trace:
        /_/src/Shared/E2ETesting/TemplatePackageInstaller.cs(54,0): at Templates.Test.Helpers.TemplatePackageInstaller.EnsureTemplatingEngineInitializedAsync(ITestOutputHelper output)
        /_/src/Shared/E2ETesting/ProjectFactoryFixture.cs(40,0): at Templates.Test.Helpers.ProjectFactoryFixture.GetOrCreateProject(String projectKey, ITestOutputHelper output)
        /_/src/ProjectTemplates/test/RazorClassLibraryTemplateTest.cs(28,0): at Templates.Test.RazorClassLibraryTemplateTest.RazorClassLibraryTemplate_WithViews_Async()
        --- End of stack trace from previous location ---
        ----- Inner Stack Trace -----
           at System.Linq.ThrowHelper.ThrowNoMatchException()
           at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
        /_/src/Shared/E2ETesting/TemplatePackageInstaller.cs(46,0): at Templates.Test.Helpers.TemplatePackageInstaller..cctor()
```

I see this in the following logs:
- Linux: https://dev.azure.com/dnceng/_apis/resources/Containers/4029389/Linux_Test_Logs?itemPath=Linux_Test_Logs%2FRelease%2FQuarantined%2FProjectTemplates.Tests_netcoreapp5.0_x64.log
- macOS: https://dev.azure.com/dnceng/_apis/resources/Containers/4029389/MacOS_Test_Logs?itemPath=MacOS_Test_Logs%2FRelease%2FQuarantined%2FProjectTemplates.Tests_netcoreapp5.0_x64.log
- Windows: https://dev.azure.com/dnceng/_apis/resources/Containers/4029389/Windows_Test_Logs?itemPath=Windows_Test_Logs%2FRelease%2FQuarantined%2FProjectTemplates.Tests_netcoreapp5.0_x64.log

The following code is `throw`ing:
``` c#
        public static string CustomHivePath { get; } = (string.IsNullOrEmpty(Environment.GetEnvironmentVariable("helix"))) 
                    ? typeof(TemplatePackageInstaller)
                        .Assembly.GetCustomAttributes<AssemblyMetadataAttribute>()
                        .Single(s => s.Key == "CustomTemplateHivePath").Value
                    : Path.Combine("Hives", ".templateEngine");
```

I'm building locally to check whether the test assembly has the expected attributes.

----

Worse, the quarantined macOS test run times out fairly consistently:
```
2020-05-04T01:55:07.6962090Z   Running tests: /Users/runner/runners/2.166.4/work/1/s/src/ProjectTemplates/test/bin/Release/netcoreapp5.0/ProjectTemplates.Tests.dll [netcoreapp5.0|x64]
2020-05-04T01:55:08.1424990Z XUnit : error : Tests failed: /Users/runner/runners/2.166.4/work/1/s/artifacts/TestResults/Release/Quarantined/BlazorTemplates.Tests_netcoreapp5.0_x64.html [netcoreapp5.0|x64] [/Users/runner/runners/2.166.4/work/1/s/src/ProjectTemplates/BlazorTemplates.Tests/BlazorTemplates.Tests.csproj]
2020-05-04T01:55:08.1518620Z ##[error]XUnit(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Test) Tests failed: /Users/runner/runners/2.166.4/work/1/s/artifacts/TestResults/Release/Quarantined/BlazorTemplates.Tests_netcoreapp5.0_x64.html [netcoreapp5.0|x64]
2020-05-04T01:55:08.8011390Z XUnit : error : Tests failed: /Users/runner/runners/2.166.4/work/1/s/artifacts/TestResults/Release/Quarantined/ProjectTemplates.Tests_netcoreapp5.0_x64.html [netcoreapp5.0|x64] [/Users/runner/runners/2.166.4/work/1/s/src/ProjectTemplates/test/ProjectTemplates.Tests.csproj]
2020-05-04T01:55:08.8018250Z ##[error]XUnit(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Test) Tests failed: /Users/runner/runners/2.166.4/work/1/s/artifacts/TestResults/Release/Quarantined/ProjectTemplates.Tests_netcoreapp5.0_x64.html [netcoreapp5.0|x64]
2020-05-04T01:56:13.3078640Z   Tests succeeded: /Users/runner/runners/2.166.4/work/1/s/artifacts/bin/dotnet-watch.Tests/Release/netcoreapp5.0/Microsoft.DotNet.Watcher.Tools.Tests.dll [netcoreapp5.0|x64]
2020-05-04T04:36:07.4376220Z ##[error]The operation was canceled.
2020-05-04T04:36:07.4382080Z ##[section]Finishing: Run Quarantined Tests
```

In the last three failed validation runs, it's the Libuv.BindTests run that never completes. But https://dev.azure.com/dnceng/_apis/resources/Containers/4029389/MacOS_Test_Logs?itemPath=MacOS_Test_Logs%2FRelease%2FQuarantined%2FLibuv.BindTests_netcoreapp5.0_x64.log is truncated and no errors show up. I can't tell whether we should be skipping those quarantined tests or fixing another build issue.

@davidfowl could you please run your tools on https://dev.azure.com/dnceng/public/_build/results?buildId=629350 ? I'm hoping my manual checks missed something.

dougbu commented:
@anurse @BrennanConroy @jkotalik and @HaoK the LibUV.BindTests have been skipped in Helix runs since #8045 went in and all of you commented on the related #8046 issue. Have these tests been reliable in our non-Helix macOS runs?

What do you recommend to find the reason for the timeouts with a newer SDK? For example, if one of you has a macOS system to test with locally, please sing out.

BrennanConroy commented:
>A prompt might appear to ask for permission to access the key. When that happens, select 'Always Allow' to grant 'dotnet' access to the certificate key in the future.

Maybe there is a prompt on the machine which is causing it to never complete.

dougbu commented:
What are you quoting @BrennanConroy❔ And, is this something you can test❔

dougbu commented:
The quarantined Blazor and project template tests are failing because `@(AssemblyAttribute)` items in the `PrepareForTest` targets of both ProjectTemplates.Tests.csproj do not run before the `CoreCompile` target runs. Interestingly, some `BeforeTargets="BeforeCompile"` _are_ working properly e.g. those in E2ETesting.targets. I'll look at the preprocessed output from `msbuild`&hellip;

BrennanConroy commented:
That's from the test logs. Although, looks like that same log appears for successful runs too.

I can try running on my Mac and see if it repros.

dougbu commented:
On the failing quarantined tests: Problem was an attempt to add attributes too late. We were just lucky `BeforeTargets="CoreCompile"` _ever_ worked. I'm switching to `BeforeTargets="GetAssemblyAttributes"` everywhere for consistency. (`BeforeTargets="BeforeCompile"` was fine but a bit of a misdirection.)

dougbu committed:
Always add `@(AssemblyAttribute)` items before they are processed
- use `BeforeTargets="GetAssemblyAttributes"` for consistently throughout the repo

dougbu commented:
:up::date: to grab latest from 'master' branch and fix the `@(AssemblyAttribute)` use in the Blazor and project template tests. Unfortunately, it's doubtful this will fix the timeouts in the macOS quarantined test runs.

BrennanConroy commented:
Same or similar issue repros locally. Does a new SDK bring a new hostfxr? I am attached to the hung process right now and it is stuck inside of native code in libcoreclr.dylib, and I noticed that there are some recent changes to that section of code in the Runtime. So I'd like to know if the host can be updated by a SDK update since this PR isn't bringing in a new Runtime.

BrennanConroy commented:
Oh, the shared runtime did update, I didn't notice that. So, this might be a regression in hostfxr.

BrennanConroy commented:
Huh, I have no idea anymore, I get the hanging tests on master branch too on my Mac.

dougbu commented:
@BrennanConroy did you delete the .dotnet/ folder before trying again in 'master'? Some parts of the build seem to roll forward when multiple SDKs are available on dev machines. This probably means `msbuild` or `dotnet` processes are reused but I'm not sure.

BrennanConroy commented:
./clean.sh deletes .dotnet/ right?

dougbu commented:
Nope, `-e .dotnet/ -e .tools/`

dougbu commented:
`git clean -dfx -e .tools/` would remove .dotnet/

BrennanConroy commented:
Oh wait, I did a manual `rm -r .dotnet`, I forgot

dougbu commented:
And, did you burn any leftover `dotnet` processes w/ fire or do the 'master' build w/ '/nr:false`?

BrennanConroy commented:
Yep (except for a zombie dotnet that just wont go away, but it's a zombie so it doesn't affect anything)

dougbu commented:
@BrennanConroy zombies 🧟 are not to be trifled with.

> I get the hanging tests on master branch too on my Mac.

More seriously, what is your Mac doing exactly when the timeouts occur❔

And, because we're not seeing similar problems in the rolling builds of our 'master' branch, I wonder about scorched Earth approaches like clearing your NuGet caches and rebooting.

dougbu commented:
/fyi latest validation build (https://dev.azure.com/dnceng/public/_build/results?buildId=630538) showed some improvement. But,
- The quarantined LibUv.BindTests run still timed out.
  - See https://dev.azure.com/dnceng/_apis/resources/Containers/4036057/MacOS_Test_Logs?itemPath=MacOS_Test_Logs%2FRelease%2FQuarantined%2FLibuv.BindTests_netcoreapp5.0_x64.log
  - @ericstj is there someone @BrennanConroy could work with to figure out if `hostfxr` has indeed regressed❔ Note, we're not sure where our build gets that executable because dotnet/runtime bits are _not_ updated in this PR.
- All (2) of the quarantined BlazorTemplates.Tests failed though they seemed to get further. New problems are certificate-related, We don't have https://whatsbroken.azurewebsites.net/ data for these tests but a quick check of recent rolling builds of 'master' didn't indicate an existing problem. But, we're taking a big jump in the SDK here.
  - See https://dev.azure.com/dnceng/_apis/resources/Containers/4036057/MacOS_Test_Logs?itemPath=MacOS_Test_Logs%2FRelease%2FQuarantined%2FBlazorTemplates.Tests_netcoreapp5.0_x64.log
- Ten of the quarantined ProjectTemplates.Tests failed in similar ways to the ProjectTemplates.Tests. This is much better than before but not great.
  - See https://dev.azure.com/dnceng/_apis/resources/Containers/4036057/MacOS_Test_Logs?itemPath=MacOS_Test_Logs%2FRelease%2FQuarantined%2FProjectTemplates.Tests_netcoreapp5.0_x64.log
  - @javiercn do you have context on recent macOS-specific changes in the SDK❔ Other ideas?

Bottom line, the remaining test failures are specific to quarantined tests but they are also consistent enough to imply we may still be dealing with product issues such as the potential `hostfxr` bug @BrennanConroy mentioned earlier.

BrennanConroy commented:
>     * if `hostfxr` has indeed regressed

I'm doubting this is the case anymore

dougbu commented:
> I'm doubting this is the case anymore

Any new and improved ideas❔

BrennanConroy commented:
Nope, my machine is unreliable since I can't even complete the tests on the master branch

javiercn commented:
@dougbu what makes you think the error has to do anything with the certificate? That error merely implies that the app is failing to start at all.

dougbu commented:
@javiercn the last server message before the client gives up is
```
The application is trying to access the ASP.NET Core developer certificate key. A prompt might appear to ask for permission to access the key. When that happens, select 'Always Allow' to grant 'dotnet' access to the certificate key in the future.
```
The server doesn't get around to emitting its usual "ready" messages. In addition, hasn't the notarization logic in the CLI and SDK changed since 5.0 preview 2?

javiercn commented:
> @javiercn the last server message before the client gives up is

Yeah, sorry. I lost track of this PR and couldn't answer. I already looked into it and there was an issue with the cert loading code. I put out a fix and improved diagnostics and I'll merge it soon.

javiercn commented:
> The server doesn't get around to emitting its usual "ready" messages. In addition, hasn't the notarization logic in the CLI and SDK changed since 5.0 preview 2?

Yes, that's why we are seeing this issue I think. The server is not picking up the certificate we generate I believe.

BrennanConroy commented:
> I already looked into it and there was an issue with the cert loading code. I put out a fix and improved diagnostics and I'll merge it soon.

Your change is only for the Templates code, however some Kestrel tests are hanging with a similar issue it looks like. They are trying to access the dev cert and never completing. Were there recent changes to how the cert code works? Do we need to do some manual cert injection for the tests in Kestrel that use https?

dougbu commented:
/fyi I'll rerun the validation builds after #21493 and #21516 are in. Will handle the #21493 merge since @javiercn needs his sleep.

dougbu commented:
/azp run

BrennanConroy commented:
Yay, Mac passed 😄

ViktorHofer commented:
🎉 🎉 🎉
